README.org
#+title: npm-transient
#+author: Lucien Cartier-Tilet
#+email: lucien@phundrak.com
[[https://github.com/Phundrak/npm-transient/actions/workflows/workflow.yml][file:https://github.com/Phundrak/npm-transient/actions/workflows/workflow.yml/badge.svg]]

* Introduction
~npm-transient~ provides a transient menu for working with npm projects.
This fork is based on @mojochao’s [[https://github.com/mojochao/npm-mode][original package]] ~npm-mode~.

* Table of Contents                                                :TOC_2_gh:
- [[#introduction][Introduction]]
- [[#installation][Installation]]
  - [[#manual-installation][Manual Installation]]
  - [[#straight-and-use-package][Straight and use-package]]
- [[#usage][Usage]]
- [[#acknowledgements][Acknowledgements]]

* Installation
This project is not yet available on any package repository, so your
only options are manual or semi-manual installation as shown below.

** Manual Installation
Start by cloning the ~npm-transient~ repository wherever you wish to:
#+begin_src sh
git clone https://github.com/Phundrak/npm-transient.git
#+end_src

Then, edit your Emacs configuration file (~$HOME/.emacs~,
~$HOME/.emacs.d/init.el~, or ~$XDG_CONFIG_DIR/emacs/init.el~) to include
the following lines:
#+begin_src emacs-lisp
(add-to-list 'load-path "path/to/your/clone/npm-transient")
(require 'npm-transient)
#+end_src

** Straight and use-package
You can use ~straight~ with ~use-package~ to install this version of
~npm-transinet~. To do so, you can use the following snippet:
#+begin_src emacs-lisp
(use-package npm-transient
  :straight (npm-transient :type git
                           :host github
                           :fork "Phundrak/npm-transient"))
#+end_src

* Usage
The transient menu can be called with ~M-x npm-transient~.

* Acknowledgements
This repo is a rewrite of @mojochao’s [[https://github.com/mojochao/npm-mode][original package]], and its
history has been preserved.


CHANGELOG.org
#+title: Changelog

* v0.7.0
- Rename ~npm-mode~ to ~npm-transient~
- Move ~CHANGES.md~ to ~CHANGELOG.org~
- Move ~TODO.md~ to ~TODOs.org~
- Move ~README.md~ to ~README.org~.
- Add transient menu ~npm-transient~
- Remove the following functions in favor of ~npm-transient~:
  - ~npm-mode-npm-init~
  - ~npm-mode-npm-install~
  - ~npm-mode-npm-install-save~
  - ~npm-mode-npm-install-save-dev~
  - ~npm-mode-npm-uninstall~
  - ~npm-mode-npm-list~
  - ~npm-mode-visit-project-file~
- Add possibility to install and uninstall non-regural packages
- Drop support for Emacs 24
- Fix readability of compilation-mode buffers

* v0.6.0
- Add ~CHANGES.md~ file to track changes.
- Add ~npm-mode-npm-list~ command and ~C-c n l~ keybinding.


npm-mode.el
;;; npm-mode.el --- minor mode for working with npm projects

;; Version: 0.6.0
;; Author: Allen Gooch <allen.gooch@gmail.com>
;; Maintainer: Lucien Cartier-Tilet <lucien@phundrak.com>
;; Url: https://github.com/Phundrak/npm-mode
;; Keywords: convenience, project, javascript, node, npm
;; Package-Requires: ((emacs "25.1"))

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; This package allows you to easily work with npm projects.  It provides
;; a minor mode for convenient interactive use of API with a
;; mode-specific command keymap.
;;
;; | command                       | keymap       | description                         |
;; |-------------------------------|--------------|-------------------------------------|
;; | npm-mode-npm-init             | <kbd>n</kbd> | Initialize new project              |
;; | npm-mode-npm-install          | <kbd>i</kbd> | Install all project dependencies    |
;; | npm-mode-npm-install-save     | <kbd>s</kbd> | Add new project dependency          |
;; | npm-mode-npm-install-save-dev | <kbd>d</kbd> | Add new project dev dependency      |
;; | npm-mode-npm-uninstall        | <kbd>u</kbd> | Remove project dependency           |
;; | npm-mode-npm-list             | <kbd>l</kbd> | List installed project dependencies |
;; | npm-mode-npm-run              | <kbd>r</kbd> | Run project script                  |
;; | npm-mode-visit-project-file   | <kbd>v</kbd> | Visit project package.json file     |
;; |                               | <kbd>?</kbd> | Display keymap commands             |

;;; Credit:

;; This package began as a fork of the emacsnpm package, and its
;; repository history has been preserved.  Many thanks to Alex
;; for his contribution.
;; https://github.com/AlexChesters/emacs-npm repo.

;;; Code:

(require 'json)
(require 'transient)

(defgroup npm-mode nil
  "Customization group for `npm-mode'."
  :prefix "npm-mode-"
  :group 'tools
  :group 'convenience
  :link '(url-link :tag "Github" "https://github.com/Phundrak/eshell-info-banner.el"))

(defcustom npm-mode-comint t
  "Colorize NPM’s buffer.
See the argument COMINT in function `compile'."
  :group 'npm-mode
  :type 'boolean
  :safe #'booleanp)

(defcustom npm-mode-command-prefix "C-c n"
  "Prefix for npm-mode."
  :group 'npm-mode
  :type 'string)

(defvar npm-mode--project-file-name "package.json"
  "The name of npm project files.")

(defvar npm-mode--modeline-name " npm"
  "Name of npm mode modeline name.")

(defun npm-mode--ensure-npm-module ()
  "Assert the current directory is inside an npm module."
  (npm-mode--project-file))

(defun npm-mode--project-file ()
  "Return path to the project file, or nil.
If project file exists in the current working directory, or a
parent directory recursively, return its path.  Otherwise, return
nil."
  (let ((dir (locate-dominating-file default-directory npm-mode--project-file-name)))
    (unless dir
      (error (concat "Error: cannot find " npm-mode--project-file-name)))
    (concat dir npm-mode--project-file-name)))

(defun npm-mode--get-project-property (prop &optional suffix)
  "Get the given PROP from the current project file.
If SUFFIX is non-nil, then it gets suffixed to the car of each
value if the result is a list."
  (let* ((project-file (npm-mode--project-file))
         (json-object-type 'hash-table)
         (json-contents (with-temp-buffer
                          (insert-file-contents project-file)
                          (buffer-string)))
         (json-hash (json-read-from-string json-contents))
         (value (gethash prop json-hash))
         values)
    (if (not (hash-table-p value))
        value
      (maphash (lambda (key value)
                 (setq values
                       (append values
                               `((,(if suffix (concat key " " suffix) key) . ,key)))))
               value)
      values)))

(defun npm-mode--get-project-scripts ()
  "Get a list of project scripts."
  (npm-mode--get-project-property "scripts"))

(defun npm-mode--get-project-dependencies ()
  "Get a list of project dependencies."
  (append (npm-mode--get-project-property "dependencies")
          (npm-mode--get-project-property "devDependencies" "(dev)")
          (npm-mode--get-project-property "peerDependencies" "(peer)")
          (npm-mode--get-project-property "bundledDependencies" "(bundled)")
          (npm-mode--get-project-property "optionalDependencies" "(optional)")))

(defun npm-mode--exec-process (cmd)
  "Execute a process running CMD."
  (let ((compilation-buffer-name-function
         (lambda (mode)
           (format "*npm:%s - %s*"
                   (npm-mode--get-project-property "name") cmd))))
    (message (concat "Running " cmd))
    (compile cmd npm-mode-comint)))

(defun npm-mode-npm-clean ()
  "Run the \\='npm list\\=' command."
  (interactive)
  (let ((dir (concat (file-name-directory (npm-mode--ensure-npm-module)) "node_modules")))
    (if (file-directory-p dir)
      (when (yes-or-no-p (format "Are you sure you wish to delete %s?" dir))
        (npm-mode--exec-process (format "rm -rf %s" dir)))
      (message (format "%s has already been cleaned" dir)))))

(defun npm-mode-npm-init ()
  "Run the \\='npm init\\=' command."
  (interactive)
  (npm-mode--exec-process "npm init"))

(defun npm-mode-npm-install ()
  "Run the \\='npm install\\=' command."
  (interactive)
  (npm-mode--exec-process "npm install"))

;;; Installing Dependencies

(defun npm-mode-npm-install-save (dep &optional dest)
  "Run the \\='npm install --save\\=' command for DEP.
If DEST is non-nil or if it has the value \\='regular, then
DEP is installed as a regular dependency. Otherwise, it is
installed as one of the following dependencies (and valid
values):
- dev
- peer
- bundled
- optional
DEST can be a string or a symbol."
  (interactive (list (read-string "Enter package name: ")
                     (completing-read "Type of dependency: "
                                      '(regular
                                        dev
                                        peer
                                        bundle
                                        optional))))
  (npm-mode--exec-process (format "npm install %s --save%s"
                                  dep
                                  (if (or (not dest)
                                          (string= dest "regular"))
                                      ""
                                    (format "-%s" dest)))))

(defun npm-mode-npm-install-save-regular (dep)
  "Run the \\='npm install --save\\=' command for DEP."
  (interactive "sEnter package name: ")
  (npm-mode-npm-install-save dep))

(defun npm-mode-npm-install-save-dev (dep)
  "Run the \\='npm install --save-dev\\=' command for DEP."
  (interactive "sEnter package name: ")
  (npm-mode-npm-install-save dep 'dev))

(defun npm-mode-npm-install-save-peer (dep)
  "Run the \\='npm install --save-peer\\=' command for DEP."
  (interactive "sEnter package name: ")
  (npm-mode-npm-install-save dep 'peer))

(defun npm-mode-npm-install-save-bundle (dep)
  "Run the \\='npm install --save-bundle\\=' command for DEP."
  (interactive "sEnter package name: ")
  (npm-mode-npm-install-save dep 'bundle))

(defun npm-mode-npm-install-save-optional (dep)
  "Run the \\='npm install --save-optional\\=' command for DEP."
  (interactive "sEnter package name: ")
  (npm-mode-npm-install-save dep 'optional))


(defun npm-mode-npm-uninstall ()
  "Run the \\='npm uninstall\\=' command."
  (interactive)
  (let* ((deps (npm-mode--get-project-dependencies))
         (dep-read (completing-read "Uninstall dependency: " deps))
         (dep (cdr (assoc-string dep-read deps))))
    (npm-mode--exec-process (format "npm uninstall %s" dep))))

(defun npm-mode-npm-list ()
  "Run the \\='npm list\\=' command."
  (interactive)
  (npm-mode--exec-process "npm list --depth=0"))

(defun npm-run--read-command ()
  "Read command from user."
  (completing-read "Run script: " (npm-mode--get-project-scripts)))

(defun npm-mode-npm-run (script)
  "Run the \\='npm run\\=' command on a project SCRIPT."
  (interactive
   (list (npm-run--read-command)))
  (npm-mode--exec-process (format "npm run %s" script)))

(defun npm-mode-visit-project-file ()
  "Visit the project file."
  (interactive)
  (find-file (npm-mode--project-file)))


;;; Transient definitions
(transient-define-prefix npm-mode--transient-add-dependency ()
  ["Install dependency as"
   ("r" "Regular dependency" npm-mode-npm-install-save-regular)
   ("d" "Dev dependency" npm-mode-npm-install-save-dev)
   ("p" "Peer dependency" npm-mode-npm-install-save-peer)
   ("b" "Bundle dependency" npm-mode-npm-install-save-bundle)
   ("o" "Optional dependency" npm-mode-npm-install-save-optional)])

(transient-define-prefix npm-mode-transient ()
  ["Dependencies"
   ("i" "Install dependencies" npm-mode-npm-install)
   ("s" "Add dependency" npm-mode--transient-add-dependency)
   ("u" "Uninstall depedency" npm-mode-npm-uninstall)
   ("l" "List dependencies" npm-mode-npm-list)]
  ["Actions"
   ("n" "Init project" npm-mode-npm-init)
   ("r" "Run script" npm-mode-npm-run)
   ("v" "Visit package.json" npm-mode-visit-project-file)
   ("q" "Quit" (lambda () (interactive) nil))])


;;; Keymap definition

(defvar npm-mode-command-keymap
  (let ((map (make-sparse-keymap)))
    (define-key map "n" 'npm-mode-npm-init)
    (define-key map "i" 'npm-mode-npm-install)
    (define-key map "s" 'npm-mode-npm-install-save)
    (define-key map "d" 'npm-mode-npm-install-save-dev)
    (define-key map "u" 'npm-mode-npm-uninstall)
    (define-key map "l" 'npm-mode-npm-list)
    (define-key map "r" 'npm-mode-npm-run)
    (define-key map "v" 'npm-mode-visit-project-file)
    (define-key map "t" 'npm-mode-transient)
    map)
  "Keymap for npm-mode commands.")

(defvar npm-mode-keymap
  (let ((map (make-sparse-keymap)))
    (define-key map (kbd npm-mode-command-prefix) npm-mode-command-keymap)
    map)
  "Keymap for `npm-mode'.")


;;; Minor mode definitions

;;;###autoload
(define-minor-mode npm-mode
  "Minor mode for working with npm projects."
  :init-value nil
  :lighter npm-mode--modeline-name
  :keymap npm-mode-keymap
  :group 'npm-mode)

;;;###autoload
(define-globalized-minor-mode npm-global-mode
  npm-mode
  npm-mode)

(provide 'npm-mode)
;;; npm-mode.el ends here


npm-transient.el
;;; npm-transient.el --- transient menu for working with npm projects

;; Version: 0.6.0
;; Author: Allen Gooch <allen.gooch@gmail.com>
;; Maintainer: Lucien Cartier-Tilet <lucien@phundrak.com>
;; Url: https://github.com/Phundrak/npm-transient
;; Keywords: convenience, project, javascript, node, npm
;; Package-Requires: ((emacs "25.1"))

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; This package allows you to easily work with npm projects.  It
;; provides a transient menu for interacting with your package
;; manager.  To launch it, simply run `M-x npm-transient'.

;;; Credit:

;; This package began as a fork of the npm-mode package, and its repository history has been preserved.

;;; Code:

(require 'json)
(require 'transient)
(require 'tabulated-list)


;;; Group, custom variables, variables, and constants

(defgroup npm-transient nil
  "Customization group for `npm-transient'."
  :prefix "npm-transient-"
  :group 'tools
  :group 'convenience
  :link '(url-link :tag "Github" "https://github.com/Phundrak/eshell-info-banner.el"))

(defcustom npm-transient-comint t
  "Colorize NPM’s buffer.
See the argument COMINT in function `compile'."
  :group 'npm-transient
  :type 'boolean
  :safe #'booleanp)

(defcustom npm-transient-command-prefix "C-c n"
  "Prefix for npm-transient."
  :group 'npm-transient
  :type 'string)

(defvar npm-transient--project-file-name "package.json"
  "The name of npm project files.")


;;; Internal transient
(eval-when-compile
  (defmacro npm-transient--define-infix (key name description type default &rest reader)
    "Define infix and its corresponding variable at once.
The variable is named npm-transient--NAME, is of type TYPE, and has a
docstring DESCRIPTION.
The KEY and READER are for the infix declaration."
    (let ((var-name (concat "npm-transient--" name)))
      `(progn
         (defcustom ,(intern var-name) ,default
           ,(concat description "
Internal variable for transient menus.")
           :type ,type
           :group 'npm-transient)
         (transient-define-infix ,(intern (concat "npm-transient--set-" name)) ()
           ,(format "Set %s option." var-name)
           :class 'transient-lisp-variable
           :variable ',(intern var-name)
           :key ,key
           :description ,description
           :argument ,(concat "--" name)
           :reader (lambda (&rest _) ,@reader)))))

  (npm-transient--define-infix
   "-p" "package-manager-arguments" "Package manager arguments"
   'string ""
   (read-string "Package manager arguments: "))

  (npm-transient--define-infix
   "-c" "command-arguments" "Package manager’s subcommand arguments"
   'string ""
   (read-string "Package manager’s subcommand arguments: "))

  (npm-transient--define-infix
   "-d" "dep-dest" "How to install the dependency"
   'symbol 'regular
   (intern (completing-read "How to install the dependency: "
                            '(regular dev peer bundle optional))))

  (npm-transient--define-infix
   "-D" "list-depth" "Depth when listing dependencies (unused for now)"
   'integer 0
   (round (read-number "Depth for listing dependencies: "))))


;;; Internal functions

(defun npm-transient--ensure-npm-module ()
  "Assert the current directory is inside an npm module."
  (npm-transient--project-file))

(defun npm-transient--project-file ()
  "Return path to the project file, or nil.
If project file exists in the current working directory, or a
parent directory recursively, return its path.  Otherwise, return
nil."
  (let ((dir (locate-dominating-file default-directory npm-transient--project-file-name)))
    (unless dir
      (error (concat "Error: cannot find " npm-transient--project-file-name)))
    (concat dir npm-transient--project-file-name)))

(defun npm-transient--get-project-property (prop &optional suffix)
  "Get the given PROP from the current project file.
If SUFFIX is non-nil, then it gets suffixed to the car of each
value if the result is a list."
  (let* ((project-file (npm-transient--project-file))
         (json-object-type 'hash-table)
         (json-contents (with-temp-buffer
                          (insert-file-contents project-file)
                          (buffer-string)))
         (json-hash (json-read-from-string json-contents))
         (value (gethash prop json-hash))
         values)
    (if (not (hash-table-p value))
        value
      (maphash (lambda (key value)
                 (setq values
                       (append values
                               `((,(if suffix (concat key " " suffix) key) . ,key)))))
               value)
      values)))

(defun npm-transient--get-project-scripts ()
  "Get a list of project scripts."
  (npm-transient--get-project-property "scripts"))

(defun npm-transient--get-project-dependencies ()
  "Get a list of project dependencies."
  (append (npm-transient--get-project-property "dependencies")
          (npm-transient--get-project-property "devDependencies" "(dev)")
          (npm-transient--get-project-property "peerDependencies" "(peer)")
          (npm-transient--get-project-property "bundledDependencies" "(bundled)")
          (npm-transient--get-project-property "optionalDependencies" "(optional)")))

(cl-defun npm-transient--make-command (&key (command "")    (arguments "")
                                            (extra-args "") double-dash-args)
  "Generate a package manager command.
The function will automatically add package manager arguments
from `npm-transient--package-manager-arguments'.

It will then add the COMMAND, ARGUMENTS to the command,
`npm-transient--command-arguments' and EXTRA-ARGS as additional
arguments to the command.

If DOUBLE-DASH-ARGS is non-nil, it will add a double dash
followed by the content of the variable."
  (replace-regexp-in-string
   " +" " "
   (concat "npm "
           (or npm-transient--package-manager-arguments "")
           " "
           command
           " "
           arguments
           " "
           (or npm-transient--command-arguments "")
           " "
           extra-args
           (if double-dash-args (concat " -- " double-dash-args) ""))))

(defun npm-transient--exec-process (cmd)
  "Execute a process running CMD."
  (let ((compilation-buffer-name-function
         (lambda (mode)
           (format "*npm:%s - %s*"
                   (npm-transient--get-project-property "name") cmd))))
    (message (concat "Running " cmd))
    (compile cmd npm-transient-comint)))

(cl-defun npm-transient--exec-command (&key (command "")    (arguments "")
                                            (extra-args "") double-dash-args)
  "Exec package manager COMMAND.
Pass `npm-transient--package-manager-arguments' as arguments to the
package manager, and `npm-transient--command-arguments' to the
command.
Optionally pass ARGUMENTS to the command such as \\='is-odd\\='
in \\='npm install is-odd\\='.
Optionally add EXTRA-ARGS as extra arguments for the command.
Optionally, pass DOUBLE-DASH-ARGS as arguments after a double
dash."
  (npm-transient--exec-process
   (npm-transient--make-command :command command       :arguments arguments
                           :extra-args extra-args :double-dash-args double-dash-args)))

(defun npm-transient--uninstall-dep (dep)
  "Uninstall dependency DEP."
  (interactive (list
                (let* ((deps (npm-transient--get-project-dependencies))
                       (dep-read (completing-read "Uninstall dependency: " deps)))
                  (cdr (assoc-string dep-read deps)))))
  (npm-transient--exec-command :command "uninstall" :arguments dep))

(defun npm-transient--install-dep (dep)
  "Install DEP with the package manager.
Also see `npm-transient--dep-dest',
`npm-transient--package-manager-arguments', and
`npm-transient--command-arguments'."
  (interactive "sDependency name: ")
  (let ((save-argument (pcase npm-transient--dep-dest
                           ('regular  "")
                           ('dev      "--save-dev")
                           ('peer     "--save-peer")
                           ('bundle   "--save-bundle")
                           ('optional "--save-optional")
                           (otherwise (error "Unknown option in npm-transient--install-deps: %s"
                                             otherwise)))))
    (npm-transient--exec-command :command "install"
                            :arguments dep
                            :extra-args save-argument)))

(defun npm-transient--install-deps ()
  "Install all dependencies from package.json."
  (interactive)
  (npm-transient--exec-command :command "install"))

(defun npm-transient--list-dependencies ()
  "List dependencies of the current project.
The dependency depth shown depends on `npm-transient--list-depth'
which can be set in the dependencies transient menu."
  (interactive)
  (let ((cmd (npm-transient--make-command :command "list"
                                     :extra-args "--depth=0"))
        (packages))
    (with-temp-buffer
      (shell-command cmd (current-buffer) (current-buffer))
      (goto-char (point-min))
      (re-search-forward "$")
      (goto-char (1+ (point)))
      (delete-region (point) (point-min))
      (goto-char (point-min))
      (save-match-data
        (while (re-search-forward
                "\\([^[:space:]]+\\)@\\([^[:space:]]+\\)[^@]*$"
                ;; XXX: Apparently (not "@") isn’t valid prior to Emacs 27
                ;; (rx (group (+ (not space)))
                ;;     "@"
                ;;     (group (+ (not space)))
                ;;     (* (not "@"))
                ;;     eol)
                nil t)
          (add-to-list 'packages
                       (list nil (vector (match-string 1) (match-string 2)))
                       t))))
    (switch-to-buffer "*npm dependencies*")
    (setq tabulated-list-format [("Dependency" 40) ("Version" 8)])
    (setq tabulated-list-entries packages)
    (message "entries: %S" tabulated-list-entries)
    (tabulated-list-init-header)
    (tabulated-list-print)))

(defun npm-transient--clean-project-root ()
  "Delete node_modules directory in root project."
  (interactive)
  (let ((dir (concat (file-name-directory (npm-transient--ensure-npm-module)) "node_modules")))
    (if (file-directory-p dir)
        (when (yes-or-no-p (format "Are you sure you wish to delete %s?" dir))
          (delete-directory dir t nil)
          (when (yes-or-no-p (format "Remove lock file?"))
            (delete-file "package-lock.json")))
      (message (format "%s has already been cleaned" dir)))))

(defun npm-transient--init-project ()
  "Initialize an npm project."
  (interactive)
  (npm-transient--exec-command :command "init"))

(defun npm-transient-run--read-command ()
  "Read command from user."
  (completing-read "Run script: " (npm-transient--get-project-scripts)))

(defun npm-transient--run-script (script)
  "Run an npm SCRIPT."
  (interactive
   (list (completing-read "Run script: " (npm-transient--get-project-scripts))))
  (npm-transient--exec-command :comand "run" :arguments script))

(defun npm-transient--visit-project-file ()
  "Visit the project file."
  (interactive)
  (find-file (npm-transient--project-file)))


;;; User interface


;;; Transient menus
(transient-define-prefix npm-transient-manage-dependencies ()
  ["Options"
   (npm-transient--set-package-manager-arguments)
   (npm-transient--set-command-arguments)
   (npm-transient--set-dep-dest)
   (npm-transient--set-list-depth)]
  ["Commands"
   ("a" "Add dependency" npm-transient--install-dep)
   ("i" "Install project dependencies" npm-transient--install-deps)
   ("u" "Uninstall depedency" npm-transient--uninstall-dep)
   ("l" "List dependencies" npm-transient--list-dependencies)])

(transient-define-prefix npm-transient ()
  ["Options"
   (npm-transient--set-package-manager-arguments)
   (npm-transient--set-command-arguments)]
  ["Actions"
   ("c" "Clean project root" npm-transient--clean-project-root)
   ("d" "Manage dependencies" npm-transient-manage-dependencies)
   ("i" "Init project" npm-transient--init-project)
   ("r" "Run script" npm-transient--run-script)
   ("v" "Visit package.json" npm-transient--visit-project-file)
   ("q" "Quit" (lambda () (interactive) nil))])

(provide 'npm-transient)
;;; npm-transient.el ends here
